core/drop_test.go

```diff
@@ -0,0 +1,346 @@
+						maximumFaultyNodesFn: c.maxFaulty,
+
+						buildProposalFn:           buildValidProposal,
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+						insertBlockFn:             nil,
+					},
+
+					&mockTransport{multicastFn: c.gossip},
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
+	}
+
+	assert.Equal(t, uint64(5), cluster.latestHeight)
+
+	offline := int(cluster.maxFaulty()) + 1
+
+	cluster.stopN(offline)
+
+	assert.Error(t, cluster.progressToHeight(2*time.Second, 10))
+	assert.Equal(t, uint64(5), cluster.latestHeight)
+
+	cluster.startN(offline)
+
+	assert.NoError(t, cluster.progressToHeight(5*time.Second, 10))
+	assert.Equal(t, uint64(10), cluster.latestHeight)
+}
+
+/*
+	Scenario:
+	1. Cluster can reach height 5
+	2. Stop MaxFaulty nodes
+	3. Cluster can still reach height 10
+*/
+func TestDropMaxFaulty(t *testing.T) {
+	t.Parallel()
+
+	cluster := newCluster(
+		5,
+		func(c *cluster) {
+			for _, node := range c.nodes {
+				node.core = NewIBFT(
+					mockLogger{},
+					&mockBackend{
+						isValidBlockFn:         isValidProposal,
+						isValidProposalHashFn:  isValidProposalHash,
+						isValidSenderFn:        nil,
+						isValidCommittedSealFn: nil,
+						isProposerFn:           c.isProposer,
+
+						idFn:                 node.addr,
+						quorumFn:             c.quorum,
+						maximumFaultyNodesFn: c.maxFaulty,
+
+						buildProposalFn:           buildValidProposal,
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+						insertBlockFn:             nil,
+					},
+
+					&mockTransport{multicastFn: c.gossip},
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
+	}
+
+	assert.Equal(t, uint64(5), cluster.latestHeight)
+
+	cluster.stopN(int(cluster.maxFaulty()))
+
+	// higher timeout due to round-robin proposer selection
+	assert.NoError(t, cluster.progressToHeight(20*time.Second, 10))
+	assert.Equal(t, uint64(10), cluster.latestHeight)
+}
+
+/*	HELPERS */
+
+var (
+	validProposal      = []byte("valid proposal")
+	validProposalHash  = []byte("valid proposal hash")
+	validCommittedSeal = []byte("valid committed seal")
+)
+
+func isValidProposal(newProposal []byte) bool {
+	return bytes.Equal(newProposal, validProposal)
+}
+
+func buildValidProposal(_ uint64) []byte {
+	return validProposal
+}
+
+func isValidProposalHash(proposal, proposalHash []byte) bool {
+	return bytes.Equal(
+		proposal,
+		validProposal,
+	) && bytes.Equal(
+		proposalHash,
+		validProposalHash,
+	)
+}
+
+type node struct {
```

Comment:
Should this and cluster be moved to a separate file?

This drop test looks to me like e2e/integration tests, better to separate them as in https://github.com/0xPolygon/pbft-consensus/tree/main/e2e

Reply:
[31fb07a](https://github.com/0xPolygon/go-ibft/pull/31/commits/31fb07a9a91e340a35a9e1545ef44d523aa93e29)